### ISSUE TITLE
Using PackageReference instead of packages.config doesn't appear to work

### ISSUE THREAD

**Issue by jonreis** [oleg-shilo/wixsharp#620]
Hello @oleg-shilo 

I know you have covered this before, but I cannot get PackageReference to work properly with Wixsharp. I always get the error below. I have tried creating new setup projects from the Wixsharp templates, however; they all seem to use Package.config.

Any hints on how to solve this issue?

Severity	Code	Description	Project	File	Line	Suppression State
Error		The "SetEnvVar" task was not found. Check the following: 1.) The name of the task in the project file is the same as the name of the task class. 2.) The task class is "public" and implements the Microsoft.Build.Framework.ITask interface. 3.) The task is correctly declared with <UsingTask> in the project file, or in the *.tasks files located in the "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin" directory.	WixSharp Setup1	C:\Users\jonre\.nuget\packages\wixsharp\1.9.5\build\WixSharp.targets	5	

Steps to reproduce:

1) Remove the nuget packages from a wixsharp project
2) Delete your obj directory
3) Re-add the nuget packages and choose the option PackageReferences instead of Packages.config
4) Attempt to build the project

![settings2](https://user-images.githubusercontent.com/332564/55441164-b8f71c80-555e-11e9-8c4b-a514a9151710.PNG)
![settings1](https://user-images.githubusercontent.com/332564/55441165-b8f71c80-555e-11e9-9ccb-d37e446b327f.PNG)

**Comment by jonreis** [comment]
@oleg-shilo I solved this by:
1) Checking SetEnvVar.dll into source control
2) Added   <UsingTask AssemblyFile="$(SolutionDir)Build\SetEnvVar.dll" TaskName="SetEnvVar" /> to my csproj file.

Could you let me know if this is the recommended approach?

Thank you!

**Comment by oleg-shilo** [comment]
Well, not exactly :o)

For quite some time `SetEnvVar.dll` was a part of the build but because it was causing some problems with packages updates it has been removed and now WixSharp package relying on no  external resources. 

If you have problems with SetEnvVar.dll then most likely you are using older VS templates. Updating WixSharp project templetes extension will solve this problem. However... I just found that NuGet restore does not recreate the project structure the way it supposed to. 

I am changing theist issue to BUG and will investigate it.

**Comment by oleg-shilo** [comment]
After some further digging I found that NuGet package is still invoking `SetEnvVar` MSBuild target. Thus fixing templates will only solve part of the problem.

The packages have been fixed (v1.9.5.1) and published. This will solve the problem of installing package on a brand new plain .NET project.

The templates fix is yet to come.

**Comment by jonreis** [comment]
@oleg-shilo as always thank you very much for your amazing library and help using it!

**Comment by oleg-shilo** [comment]
The VS project templates extension has also been fixed updated (v1.8.0):
https://marketplace.visualstudio.com/items?itemName=OlegShilo.WixSharpProjectTemplates

Though it may take a bit of time until VSMarketplace updates the extension index API.

**Comment by jonreis** [comment]
Hi @oleg-shilo, I tried out the new templates (v1.8.0) and they are still using packages.config and placing dlls into the local project "packages" directory.  However; if you use the new PackageReference mechanism the nuget packages get stored in the <user>/.nuget folder. 

It does seem to work for me if I use the PackageReference approach and only include (added by nuget automatically) in my csproj file.

```
    <PackageReference Include="WixSharp">
      <Version>1.9.5</Version>
    </PackageReference>
```

If I look in my obj directory there is a targets file that is automatically generated by Visual Studio. It contains:

```
<?xml version="1.0" encoding="utf-8" standalone="no"?>
<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
  </PropertyGroup>
  <ImportGroup Condition=" '$(ExcludeRestorePackageImports)' != 'true' ">
    <Import Project="$(NuGetPackageRoot)wixsharp.bin\1.9.5.1\build\WixSharp.bin.targets" Condition="Exists('$(NuGetPackageRoot)wixsharp.bin\1.9.5.1\build\WixSharp.bin.targets')" />
    <Import Project="$(NuGetPackageRoot)wixsharp\1.9.5.1\build\WixSharp.targets" Condition="Exists('$(NuGetPackageRoot)wixsharp\1.9.5.1\build\WixSharp.targets')" />
  </ImportGroup>
</Project>
```

So to summarize, every things seems work without having to use the WixSharpProjectTemplates if you use the PackageReference approach. Are the templates for backward compatibility, or am I missing something? What is the minimum requirement in the csproj file for Wixsharp to function properly using the PackageReference approach?

**Comment by jonreis** [comment]
@oleg-shilo I think I must have closed this accidentally. Please see my comment above.

**Comment by oleg-shilo** [comment]
I think there is a bit of confusion about this issue. I need to explain....

The problem with `*targets` reference to _SetEnvVar.dll_ was areal problem and it is solved now.

Having nuget references as a `packages.config` is not a problem even though referencing the global packages from the project file is indeed a preferred approach.

The real problem is that having project-listed packages does not play well with project templates. Despite my all effort was not able to find the solution when the template wold allow project-listed packages. Furthermore, having project-listed packages prevents templates from being integrated with VS2015.

However if one starts with the plain C# console app then he can add nuget packages either way, globally or per project (_packages.config_)

**Comment by jonreis** [comment]
Ok, thanks for the explanation. Everything is working fine for me so I will close the bug.

**Comment by allknower** [comment]
Hello @oleg-shilo,
Can you explain, please, what is the right way to pass VS variables to wixsharp script now? I used SetEnvVar task before, but it was deleted from WixSharp package. I read comments in this issue and in https://github.com/oleg-shilo/wixsharp/issues/569 and? of course I read [wiki](https://github.com/oleg-shilo/wixsharp/wiki/Developer's-Guide#building-environment-integration), but couldn't realize what should I do. 

Thanks!

**Comment by oleg-shilo** [comment]
`SetEnvVar` was initially introduced to pass the location o the project folder. Though later it became apparent that it created more problems than bringing the benefits whan upgrading the package.  Particularly because the location of the project folder may also be safely determined from the _current dir_ during the build. That's why it was eventually deleted from the solution.

But if you want to pass any sort of information into the project you can still red envars that you may set in the pre-build project events.

But if you want a complete flexibility then you can install `WixSharp.bin` package instead of `WixSharp`. This way tere will be no MSBuild custom task and building msi will need to be triggered from the post-build event by running the `$(TargetPath)` executable. And when you d that you can always pass any app arguments.

**Comment by smaudet** [comment]
Hmm. I'm also trying to use SetEnvVars. @oleg-shilo Could you possibly furnish the commit where it was deleted?

My use case is different - I have multiple project configurations and I would like to access those. I can't seem to find this feature in wixsharp, this seems the closest. However, it doesn't seem to work... (could be an msbuild/dotnet mismatch).

All we have is the dll and the source file, with no way to build. Hard to debug...

**Comment by oleg-shilo** [comment]
`MSBuild.SetEnvVariable` is no longer used. Bot streams six4 and wix3 are relying on a completely reworked build process that uses project references (at least wix4).

If you are creating a new project use VS templates.
If you are upgrading the project then create a new project from the template and update the project file manually

**Comment by smaudet** [comment]
Yes I ended up impling my own msbuild task, there was no way to pull build config settings from the csproj.

SetEnvVariable wasn't any good either as env vars are transient.

I used the VS templates but those provided no way to pull build information.

**Comment by oleg-shilo** [comment]
> no way to pull build information.

Can you elaborate on that? Not sure what you are trying to achieve.

**Comment by smaudet** [comment]
There is various information available to the build system, most notably $(Configuration), but also any and all build-related information such as package reference paths, etc. (e.g. for pulling/using nuget stuff).

So I have a different build and different packaging needs for e.g. Debug vs Release (perhaps not a common need, but one I want nonetheless).

**Comment by oleg-shilo** [comment]
Currently the old `SetEnvarVariable` is replaced with the shell command in the project file:

```xml
  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="cd .\&#xD;&#xA;set ide=true&#xD;&#xA;&quot;$(TargetPath)&quot;" />
  </Target>
```

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/0c50858b-4456-4607-9eea-a26349f647d1)

You can extend this approach to do more, like passing the context for the build script:

![image](https://github.com/oleg-shilo/wixsharp/assets/16729806/78c251bb-12e8-4599-89e0-4985fc488b27)

And then:

```C#
static void Main(string[] args)
{
    var myArg = args.FirstOrDefault(x=>x.StartsWith("-myArg"))?.Split(':').Last();
```

**Comment by smaudet** [comment]
Ah, that's what that thing does. I have it commented out, it was causing build failures, and the unicode escapes don't make for easy reading.

**Comment by oleg-shilo** [comment]
These two link can help to understand the build process:

https://github.com/oleg-shilo/wixsharp/wiki/VS2019-%E2%80%93-2022-Templates

![image](https://private-user-images.githubusercontent.com/16729806/278498389-2e287633-6c49-424d-acc9-2862730234dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ2MTY0MjQsIm5iZiI6MTcxNDYxNjEyNCwicGF0aCI6Ii8xNjcyOTgwNi8yNzg0OTgzODktMmUyODc2MzMtNmM0OS00MjRkLWFjYzktMjg2MjczMDIzNGRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTAyVDAyMTUyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg5YzczZTQxNmZlYmZhOWRlNDJjODJmMGM4NzE3ZjEwODBmMDliZDU1NDhiY2ZiYzBjZjA2YjE2MjRiNGE0YWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.E82a2Gb7w1UzI11r1IJvua0PmlsODKD8ynWw-1kIa5I)

**Comment by smaudet** [comment]
Hmm. I was going to say the reason it breaks is the first time you build the target doesn't exist, I'll try it again and see what's wrong with your template.

I do think clearer documentation could be helpful, too (such as removing the old invalid hint and updating with current supported process).

**Comment by oleg-shilo** [comment]
> I do think clearer documentation could be helpful,..

I am happy to process any feedback. But I may need your help in understanding the gaps.

1. Can you please provide the details on the "old invalid hint"?

2. Another question. Where you (as a user) would expect the description of the current process to be?
I placed it in the [VS integration](https://github.com/oleg-shilo/wixsharp/wiki/VS2019-%E2%80%93-2022-Templates) section. Do you think there is a better place for it? So it is easier found.

3. And the last one. "the reason it breaks is the first time you build the target doesn't exist".
    I just retested it again and the fresh new project (from the VS template) builds right away. Is it some specific template that does not build on the first attempt? Or maybe it is something in the project content that triggered the problem in your environment? Can you elaborate on this?

**Comment by smaudet** [comment]
1. https://github.com/oleg-shilo/wixsharp/wiki/Developer's-Guide 

search for this paragraph:

`You can also map VS project variables to environment variables by using custom MSBuild task SetEnvVar.dll distributed with Wix# NuGet packages. The following XML definition placed in WixSharp*.targets file will always map the VS project variables to the environment ones so your build script can access them:`

<SetEnvVar Values="ProjectName=$(ProjectName);ProjectDir=$(ProjectDir);SolutionName=$(SolutionName);SolutionDir=$(SolutionDir)"/>

2. I'd expect it in the wiki, but I'd expect to be called out "how to pull build information" or some-such.

Your VS integration page is actually not very helpful, I found the templates without finding your page and created a project.

Then your page just looks like you're walking someone through how to use VS templates, which I suppose if you've never created a project before might be helpful, but its not what I'm looking for when I already have a working project that builds in Visual Studio (except for one weird bit of code that is a bit inscrutable).

I think the problem is you tried to make it too seamless and so when something breaks, its hidden inside weird-looking code buried at the bottom of a page that looks like it is not relevant. You show some images of setting command line parameters, but there's never any textual mention of what to do, so I never found it (especially by search).

3. I retested as well. Works for me now. With my project that didn't work, it still fails if I re-enable (works fine without). I think it may have been an issue with code I modified and not expecting it to fail during build.

Which I think is the *other* part of this - that task should perhaps be disabled by default in the build scripts, rather than on and always running (IIRC I just wanted it to build and it failed). Then you can explicitly mention how to run the installer as part of the build by uncommenting the task.

Even just a comment explaining what the task does in the template would have helped a lot...

**Comment by oleg-shilo** [comment]
Thank you @smaudet. I have updated the documentation according to your feedback.
 
 1. I have completely removed the description of the obsolete build process and gave a brief description of the [two-step build flow](https://github.com/oleg-shilo/wixsharp/wiki/Developer's-Guide#compiling-wix-project).
 
  2.  I am not sure "how to pull build information" is an intuitive title so I simply covered this topic in this VS integration page at the end of the [section](https://github.com/oleg-shilo/wixsharp/wiki/VS2019-%E2%80%93-2022-Templates#post-build-event-and-build-context).
  
> that task should perhaps be disabled by default in the build scripts.....

Unfortunately, it's not that simple. Disabling the post-build event simply prevents msi from building. Thus the build script will be compiled but never executed. And the user feedback I received previously indicated the strong expectations of msi file being produced as the result of building the VS project. Particularly, if the project is part of the CI.
